testy/runtime.py

```diff
@@ -1,25 +1,103 @@
-    studentCode.main(state)
-    nextCall += 1
-    time.sleep(nextCall - time.time())
+  badThingsQueue = multiprocessing.Queue()
+  stateQueue = multiprocessing.Queue()
+  spawnProcess = processFactory(badThingsQueue, stateQueue)
+  restartCount = 0
+  try:
+    spawnProcess(PROCESS_NAMES.STATE_MANAGER, startStateManager)
+    while True:
+      if restartCount >= 5:
+        print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+        print("Too many restarts, terminating")
+        break
+      print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+      print("Starting studentCode attempt: %s" % (restartCount,))
+      spawnProcess(PROCESS_NAMES.STUDENT_CODE, runStudentCode)
+      while True:
+        globalBadThing = badThingsQueue.get(block=True)
+        print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+        print(globalBadThing)
+        if (globalBadThing.event == BAD_EVENTS.STUDENT_CODE_ERROR) or \
+            (globalBadThing.event == BAD_EVENTS.STUDENT_CODE_TIMEOUT):
+          break
+      stateQueue.put([SM_COMMANDS.RESET])
+      restartCount += 1
+  except:
+    print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+    print("Funtime Runtime Had Too Much Fun")
+    print(traceback.print_exception(*sys.exc_info()))
+
+def runStudentCode(badThingsQueue, stateQueue, pipe):
+  try:
+    import signal
+    def timed_out_handler(signum, frame):
+      raise TimeoutError("studentCode timed out")
+    signal.signal(signal.SIGALRM, timed_out_handler)
+
+    import studentCode
+
+    r = studentAPI.Robot(stateQueue, pipe)
+    setattr(studentCode, 'Robot', r)
+
+    signal.alarm(RUNTIME_INFO.STUDENT_CODE_TIMEOUT)
+    studentCode.setup(pipe)
+    signal.alarm(0)
+
+    nextCall = time.time()
+    while True:
+      signal.alarm(RUNTIME_INFO.STUDENT_CODE_TIMEOUT)
+      studentCode.main(stateQueue, pipe)
+      signal.alarm(0)
+      nextCall += 1.0/RUNTIME_INFO.STUDENT_CODE_HZ.value
+      time.sleep(nextCall - time.time())
+  except TimeoutError:
+    badThingsQueue.put(BadThing(sys.exc_info(), None, event=BAD_EVENTS.STUDENT_CODE_TIMEOUT))
+  except Exception:
+    badThingsQueue.put(BadThing(sys.exc_info(), None, event=BAD_EVENTS.STUDENT_CODE_ERROR))
+
+def startStateManager(badThingsQueue, stateQueue, runtimePipe):
+  try:
+    SM = stateManager.StateManager(badThingsQueue, stateQueue, runtimePipe)
+    SM.start()
+  except Exception:
+    badThingsQueue.put(BadThing(sys.exc_info(), None))
+
+def processFactory(badThingsQueue, stateQueue):
+  def spawnProcessHelper(processName, helper):
+    pipeToChild, pipeFromChild = multiprocessing.Pipe()
+    if processName != PROCESS_NAMES.STATE_MANAGER:
+      stateQueue.put([SM_COMMANDS.ADD, processName, pipeToChild], block=True)
+    newProcess = multiprocessing.Process(target=helper, name=processName.value, args=(badThingsQueue, stateQueue, pipeFromChild))
+    allProcesses[processName] = newProcess
+    newProcess.daemon = True
+    newProcess.start()
+  return spawnProcessHelper
 
 runtime()
```

Comment:
Might want to do

```python
if __name__ == "__main__":
    runtime()
```

(this is the canonical way to run a function if a python file is run directly, but not if it's improted)